File: src/04-challenge/solution.js
export default function fn() {
	const pin = []

	const blocks = code.replaceAll('[', '').split(']')
	blocks.pop()

	const actions = {
		'+': 1,
		'-': -1,
	}

	for (const block of blocks) {
		const number = block[0]

		if (number === '<') {
			pin.length && pin.push(pin[pin.length - 1])
			continue
		}

		let pinNumber = parseInt(number)
		const operations = block.replaceAll(/\d/g, '')

		for (const operation of operations) {
			pinNumber += actions[operation]
		}

		pin.push((pinNumber + 10) % 10)
	}

	return pin.length < 4 ? null : pin.join('')
}

File: src/04-challenge/solution.py
def fn(code: str) -> str | None:
    pin: list[str] = []
    blocks = code.replace("[", "").split("]")[:-1]

    actions = {
        "+": 1,
        "-": -1,
    }

    for block in blocks:
        number = block[0]

        if number == "<":
            if pin:
                pin.append(pin[-1])
            continue

        pin_number = int(number)
        operations = block[1:]

        for operation in operations:
            pin_number += actions[operation]

        pin.append(f"{(pin_number + 10) % 10}")

    return None if len(pin) < 4 else "".join(pin)
